wasm/test_page/js/worker.js

```diff
@@ -23,13 +23,68 @@ const fileInfo = [
 const encoder = new TextEncoder(); // string to utf-8 converter
 const decoder = new TextDecoder(); // utf-8 to string converter
 
+const OPTIMIZED_GEMM = "mozIntGemm";
+
+const FALLBACK_GEMM = "asm";
+  
+const GEMM_TO_FALLBACK_FUNCTIONS_MAP = {
+    "int8_prepare_a": "int8PrepareAFallback",
+    "int8_prepare_b": "int8PrepareBFallback",
+    "int8_prepare_b_from_transposed": "int8PrepareBFromTransposedFallback",
+    "int8_prepare_b_from_quantized_transposed": "int8PrepareBFromQuantizedTransposedFallback",
+    "int8_prepare_bias": "int8PrepareBiasFallback",
+    "int8_multiply_and_add_bias": "int8MultiplyAndAddBiasFallback",
+    "int8_select_columns_of_b": "int8SelectColumnsOfBFallback"
+};
+
+function createOptimizedGemm(Module, {env}) {
+  const instance = new WebAssembly.Instance(WebAssembly[OPTIMIZED_GEMM](), {"": env});
+  for (const key of Object.keys(GEMM_TO_FALLBACK_FUNCTIONS_MAP)) {
+    if (!instance.exports[key])
+      throw new Error(`Missing ${key} in ${OPTIMIZED_GEMM}`);
+  }
+  return instance.exports;
+}
+
+function createFallbackGemm(Module) {
+  const exports = {};
+  // Map function pointers to other functions that will be available once
+  // the module is loaded.
+  for (const [key, fallback] of Object.entries(GEMM_TO_FALLBACK_FUNCTIONS_MAP)) {
+    exports[key] = (...args) => Module[FALLBACK_GEMM][fallback](...args);
+  }
+  return exports;
+}
+
+function createWasmInstance(Module, info) {
+  const response = await fetch("bergamot-translator-worker.wasm");
+  const wasmBinary = await response.arrayBuffer();
+  try {
+    info['wasm_gemm'] = this.createExternalGemm(Module, info);
+    console.info(`Using gemm from ${OPTIMIZED_GEMM}`);
+  } catch (e) {
+    info['wasm_gemm'] = this.createFallbackGemm(Module, info);
+    console.info(`Using gemm from ${FALLBACK_GEMM} because`, e);
+  }
+
+  const binary = new Uint8Array(wasmBinary);
+  const {instance} = await WebAssembly.instantiate(binary, info);
+  
+  return instance;
+}
+
```

Comment:
One general comment. In my opinion, this code should still be part of the generated JS file because:

- It is common code that will be replicated for every client
- The client now needs to be aware of the optimized gemm and bears the responsibility of writing this code
- Without this code, the generated artifacts will not work out of box because even if optimized gemm is not available

If I am missing something here then please let me know. Else, I recommend to revert this specific change.